### ISSUE TITLE
Trainable, Non-trainable parameters

### ISSUE THREAD

**Issue by pharringtonp19** [facebookresearch/functorch#64]
Is it possible to perform the inner loop of MAML on only a subset of the parameters? 

[haiku](https://dm-haiku.readthedocs.io/en/latest/notebooks/non_trainable.html) in JAX makes this possible  by allowing one to split the parameters into trainable and nontrainable subsets.

**Comment by zou3519** [comment]
The short answer is yes. The long answer is, it's a little difficult to do so but it can definitely be done. We (functorch) should invest in a better API to actually express this (and we'd love your feedback on what that API should look like).

Here's how to perform the inner loop of MAML on only a subset of the parameters. Using one of our [MAML examples](https://github.com/facebookresearch/functorch/blob/7fb66ef233ef46fa9831b5458c58e24e1f55f0fd/examples/maml_regression/evjang_transforms_module.py#L75-L85) as an example, the idea is to manually partition the set of params into trainable and non-trainable:

```py
    def get_loss_for_task(x1, y1, x2, y2):
        def inner_loss(trainable_params, nontrainable_params, x1, y1):
            params = merge_params(trainable_params, nontrainable_params)
            f = net(params, x1)
            loss = mse_loss(f, y1)
            return loss

        grads = grad(inner_loss)(trainable_params, non_trainable_params, x1, y1)
        new_trainable_params = [(trainable_params[i] - alpha*grads[i]) for i in range(len(trainable_params))]

        new_params = merge_params(new_trainable_params, nontrainable_params)
        v_f = net(new_params, x2)
        return mse_loss(v_f, y2)
```
the user can define `partition_params(List) -> Tuple[List, List]` and `merge_params`:
```py
def partition_params(params):
        trainable_params = [params[0], params[2], ...]
        nontrainable_params = [params[1], ...]

def merge_params(trainable_params, non_trainable_params):
        params = [trainable_params[0], nontrainable_params[1], trainable_params[1], ...]

```

Obviously this isn't very nice and can be tedious so we are planning on figuring out a better way to do this.

**Comment by pharringtonp19** [comment]
Just an update, in Flax, it seems that the preferred solution is to actually zero out the gradients for the parameters that we don't want to update. 

The question was initially asked [here](https://github.com/google/flax/discussions/1453) and I believe that the correct answer was actually addressed in an issue in [Optax](https://github.com/deepmind/optax/issues/159)
